src/xenia/apu/xma_context.cc

```diff
@@ -806,25 +807,45 @@ int XmaContext::PrepareDecoder(uint8_t* packet, int sample_rate, int channels) {
 }
 
 bool XmaContext::ConvertFrame(const uint8_t** samples, int num_channels,
-                              int num_samples, uint8_t* output_buffer) {
+                              uint8_t* output_buffer) {
   // Loop through every sample, convert and drop it into the output array.
   // If more than one channel, we need to interleave the samples from each
   // channel next to each other.
-  // TODO: This can definitely be optimized with AVX/SSE intrinsics!
-  uint32_t o = 0;
-  for (int i = 0; i < num_samples; i++) {
-    for (int j = 0; j < num_channels; j++) {
-      // Select the appropriate array based on the current channel.
-      auto sample_array = reinterpret_cast<const float*>(samples[j]);
-
-      // Raw sample should be within [-1, 1].
-      // Clamp it, just in case.
-      float raw_sample = xe::saturate(sample_array[i]);
-
-      // Convert the sample and output it in big endian.
-      float scaled_sample = raw_sample * ((1 << 15) - 1);
-      int sample = static_cast<int>(scaled_sample);
-      xe::store_and_swap<uint16_t>(&output_buffer[o++ * 2], sample & 0xFFFF);
+  constexpr float scale = (1 << 15) - 1;
+  auto out = reinterpret_cast<int16_t*>(output_buffer);
+
+#if XE_ARCH_AMD64
+  static_assert(kSamplesPerFrame % 4 == 0);
+  // Most audio is single channel, no need to optimize for the game music
+  if (num_channels == 1) {
+    const auto in = reinterpret_cast<const float*>(samples[0]);
+    const __m128 scale_mm = _mm_set1_ps(scale);
+    // effectively packs int32 to int16 and byte swaps
+    const __m128i shufmask =
+        _mm_set_epi8(-1, -1, -1, -1, -1, -1, -1, -1, 12, 13, 8, 9, 4, 5, 0, 1);
+    for (int i = 0; i < kSamplesPerFrame; i += 4) {
+      __m128 in_mm = _mm_loadu_ps(&in[i]);
+      in_mm = _mm_mul_ps(in_mm, scale_mm);
+      __m128i out_mm = _mm_cvtps_epi32(in_mm);
+      out_mm = _mm_shuffle_epi8(out_mm, shufmask);
+      xe::store<int64_t>(&out[i], _mm_extract_epi64(out_mm, 0));
+    }
+  } else {
+#else
+  {
+#endif
+    uint32_t o = 0;
+    for (int i = 0; i < kSamplesPerFrame; i++) {
+      for (int j = 0; j < num_channels; j++) {
+        // Select the appropriate array based on the current channel.
+        auto in = reinterpret_cast<const float*>(samples[j]);
+
+        // Raw sample is within [-1, 1].
```

Comment:
Is this always true? If there's 32768 somehow, it will become -32768 in both the scalar and the AVX versions.

Reply:
fixed